clientapi/auth/password.go

```diff
@@ -62,6 +70,101 @@ func (t *LoginTypePassword) Login(ctx context.Context, req interface{}) (*Login,
 			JSON: jsonerror.InvalidUsername(err.Error()),
 		}
 	}
+	if len(t.Config.LDAP.Host) > 0 {
+		addr := ""
+		if t.Config.LDAP.TLS {
+			addr = "ldaps://" + t.Config.LDAP.Host + ":" + t.Config.LDAP.Port
+		} else {
+			addr = "ldap://" + t.Config.LDAP.Host + ":" + t.Config.LDAP.Port
+		}
+
+		var conn *ldap.Conn
+		conn, err = ldap.DialURL(addr)
+		if err != nil {
+			return nil, &util.JSONResponse{
+				Code: http.StatusUnauthorized,
+				JSON: jsonerror.InvalidUsername(err.Error()),
+			}
+		}
+		defer conn.Close()
+
+		e1 := conn.Bind(t.Config.LDAP.BindDN, t.Config.LDAP.BindPSWD)
+		if e1 != nil {
+			return nil, &util.JSONResponse{
+				Code: http.StatusUnauthorized,
+				JSON: jsonerror.InvalidUsername(err.Error()),
+			}
+		}
+		filter := fmt.Sprintf("(&%s(%s=%s))", t.Config.LDAP.Filter, "uid", localpart)
+		searchRequest := ldap.NewSearchRequest(t.Config.LDAP.BaseDN, ldap.ScopeWholeSubtree, ldap.NeverDerefAliases, 0, 0, false, filter, []string{"uid"}, nil)
+		var sr *ldap.SearchResult
+		sr, err = conn.Search(searchRequest)
+		if err != nil {
+			return nil, &util.JSONResponse{
+				Code: http.StatusUnauthorized,
+				JSON: jsonerror.InvalidUsername(err.Error()),
+			}
+		}
+		if len(sr.Entries) > 1 {
+			return nil, &util.JSONResponse{
+				Code: http.StatusUnauthorized,
+				JSON: jsonerror.BadJSON("'user' must be duplicated."),
+			}
+		}
+		if len(sr.Entries) == 0 {
+			_, err = t.GetAccountByPassword(ctx, localpart, r.Password)
+			if err != nil {
+				return nil, &util.JSONResponse{
+					Code: http.StatusForbidden,
+					JSON: jsonerror.Forbidden("username or password was incorrect, or the account does not exist"),
+				}
+			}
+			return &r.Login, nil
+		}
+
+		userDN := sr.Entries[0].DN
+		err = conn.Bind(userDN, r.Password)
+		if err != nil {
+			return nil, &util.JSONResponse{
+				Code: http.StatusUnauthorized,
+				JSON: jsonerror.InvalidUsername(err.Error()),
+			}
+		}
+
+		_, err = t.GetAccountByLocalpart(ctx, localpart)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				var accRes api.PerformAccountCreationResponse
+				err = t.UserAPI.PerformAccountCreation(ctx, &api.PerformAccountCreationRequest{
+					AppServiceID: "",
+					Localpart:    localpart,
+					Password:     r.Password,
```

Comment:
This looks like we replicate the user's LDAP password from the point of their first login into the Dendrite account database, which I don't think we want to do since we should probably already be asking LDAP each time the user authenticates with their *current* password.